File: app/src/main/java/com/udacity/google/popularmovies/fragments/DetailFragment.java
package com.udacity.google.popularmovies.fragments;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.MenuItemCompat;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.ListView;
import android.support.v7.widget.ShareActionProvider;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.ToggleButton;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.squareup.picasso.Picasso;
import com.udacity.google.popularmovies.util.FavouriteDataSource;
import com.udacity.google.popularmovies.R;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Scanner;

/**
 * Created by Dell on 10/4/2015.
 */
public class DetailFragment extends Fragment {

    int clickedMovieId;
    ListView reviewsLV;
    ListView trailersLv;
    RequestQueue queue;
    ArrayList<String> trailersNames;
    ArrayList<String> trailersKeys;
    private ShareActionProvider mShareActionProvider;
    Intent shareIntent;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        View rootView = inflater.inflate(R.layout.detail_fragment, null, false);

//
//        Scanner sc = new Scanner(System.in);
//        String s ;
//        System.out.println();

        final ToggleButton favButton = (ToggleButton) rootView.findViewById(R.id.FavtoggleButton);
        //MainFragment.clickedMovie

        clickedMovieId = MainFragment.Companion.getClickedMovie().getId();

        TextView movieTitle = (TextView) rootView.findViewById(R.id.movie_title);
        movieTitle.setText(MainFragment.Companion.getClickedMovie().getTitle());

        ImageView moviePoster = (ImageView) rootView.findViewById(R.id.movie_poster);
        Picasso.with(getActivity()).load(MainFragment.Companion.getClickedMovie().getPosterPath()).resize(300, 300).centerInside().into(moviePoster);


        Log.d("DetailDebug", MainFragment.Companion.getClickedMovie().getPosterPath());

        TextView movieDate = (TextView) rootView.findViewById(R.id.movie_date);
        movieDate.setText(MainFragment.Companion.getClickedMovie().getReleaseDate());

        TextView movieVoteAverage = (TextView) rootView.findViewById(R.id.movie_vote_average);
        movieVoteAverage.setText(MainFragment.Companion.getClickedMovie().getVoteAverage());


        TextView movieOverView = (TextView) rootView.findViewById(R.id.movie_overview);
        movieOverView.setText(MainFragment.Companion.getClickedMovie().getOverview());


        favButton.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                FavouriteDataSource favouriteDataSource = new FavouriteDataSource(getActivity());
                favouriteDataSource.open();

                if (favButton.getText().equals("MARK AS A FAVOURITE")) {

//                    save clicked movie id and poster link to db
                    favouriteDataSource.create(MainFragment.Companion.getClickedMovie());
                    Toast.makeText(getActivity(), "Marked as FAVOURITE", Toast.LENGTH_LONG).show();
                } else {

                    favouriteDataSource.delete(MainFragment.Companion.getClickedMovie().getId());
                    Toast.makeText(getActivity(), "Marked as not favourite", Toast.LENGTH_LONG).show();
                }

                favouriteDataSource.close();

            }
        });

        trailersLv = (ListView) rootView.findViewById(R.id.trailers_list_view);

        trailersLv.setOnTouchListener(new View.OnTouchListener() {
            // Setting on Touch Listener for handling the touch inside ScrollView
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                // Disallow the touch request for parent scroll on touch of child view
                v.getParent().requestDisallowInterceptTouchEvent(true);
                return false;
            }
        });


        queue = Volley.newRequestQueue(getActivity());


        trailersNames = new ArrayList<String>();
        trailersKeys = new ArrayList<String>();

        Uri.Builder builder = new Uri.Builder();

        builder.scheme("http")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath("" + clickedMovieId)
                .appendPath("videos")
                .appendQueryParameter("api_key", getActivity().getResources().getString(R.string.api_key));

//http://api.themoviedb.org/3/movie/49026/videos?api_key=4f92be250f018aff8f3a2b5c3864aecd
        String myUrl = builder.build().toString();

        Log.d("VolleyDebug", myUrl);
        JsonObjectRequest trailersRequest = new JsonObjectRequest(Request.Method.GET, myUrl, null, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {

                JSONArray resultJsonArray = null;
                try {

                    resultJsonArray = response.getJSONArray("results");


                    for (int i = 0; i < resultJsonArray.length(); i++) {

                        JSONObject jsonObject = resultJsonArray.getJSONObject(i);

                        String name = jsonObject.getString("name");
                        String key = jsonObject.getString("key");

                        trailersNames.add(name);
                        trailersKeys.add(key);
                        CustomListAdapter adapter = new CustomListAdapter(getActivity(), trailersNames);

                        trailersLv.setAdapter(adapter);

                    }

                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {
                        Toast.makeText(getActivity(), "Sorry some thing happened while retrieving trailers, please try again", Toast.LENGTH_LONG).show();

                    }
                }
        );

        queue.add(trailersRequest);


        trailersLv.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {

                String clickedKey = trailersKeys.get(i);

                String url = "https://www.youtube.com/watch?v=" + clickedKey;
                Intent intent = new Intent(Intent.ACTION_VIEW);
                intent.setData(Uri.parse(url));
                startActivity(intent);


            }
        });

        reviewsLV = (ListView) rootView.findViewById(R.id.reviews_list_view);

        final ArrayList<String> reviewsArrayList = new ArrayList<String>();


        builder = new Uri.Builder();

        builder.scheme("http")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath("" + clickedMovieId)
                .appendPath("reviews")
                .appendQueryParameter("api_key", getActivity().getResources().getString(R.string.api_key));

        //http://api.themoviedb.org/3/movie/49026/reviews?api_key=4f92be250f018aff8f3a2b5c3864aecd

        myUrl = builder.build().toString();


        JsonObjectRequest reviewsRequest = new JsonObjectRequest(Request.Method.GET, myUrl, null, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {

                try {
                    JSONArray resultsJsonArray = response.getJSONArray("results");


                    for (int i = 0; i < resultsJsonArray.length(); i++) {

                        JSONObject resultJsonObj = resultsJsonArray.getJSONObject(i);

                        String author = resultJsonObj.getString("author");

                        String content = resultJsonObj.getString("content");

                        //TODO: author with content !
                        reviewsArrayList.add(author + ": " + System.getProperty("line.separator") + System.getProperty("line.separator") + content);

                        Log.d("DetailDebug", author + ": /r/n" + content);

                        ArrayAdapter<String> reviewsAdapter = new ArrayAdapter<String>(getActivity(), android.R.layout.simple_list_item_1, reviewsArrayList);
                        reviewsLV.setAdapter(reviewsAdapter);

                    }

                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }
        },
                new Response.ErrorListener() {
                    @Override
                    public void onErrorResponse(VolleyError error) {

                    }
                }
        );

        queue.add(reviewsRequest);


        reviewsLV.setOnTouchListener(new View.OnTouchListener() {
            // Setting on Touch Listener for handling the touch inside ScrollView
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                // Disallow the touch request for parent scroll on touch of child view
                v.getParent().requestDisallowInterceptTouchEvent(true);
                return false;
            }
        });


        FavouriteDataSource favouriteDataSource = new FavouriteDataSource(getActivity());
        favouriteDataSource.open();
        if (favouriteDataSource.isFavouriteMovie(clickedMovieId)) {

            favButton.setText("FAVOURITE");
        }


        return rootView;
    }

    public class CustomListAdapter extends ArrayAdapter {

        //ArrayList<String> images ;
        Activity context;

        ArrayList<String> stringArrayList;

        public CustomListAdapter(Activity context, ArrayList<String> resources) {
            super(context, R.layout.grid_view_item);

            stringArrayList = resources;
            this.context = context;
        }

        @Override
        public int getCount() {
            return stringArrayList.size();
        }

        @Override
        public Object getItem(int position) {
            return stringArrayList.get(position);
        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {

            LayoutInflater inflater = context.getLayoutInflater();

            View rowView = inflater.inflate(R.layout.list_view_item, null, true);

            TextView tv = (TextView) rowView.findViewById(R.id.list_item_text_view);

            tv.setText(stringArrayList.get(position));

            return rowView;


        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setHasOptionsMenu(true);

    }


    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {

        inflater.inflate(R.menu.menu_detail, menu);




    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

       if(item.getItemId() == R.id.menu_item_share) {
           mShareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(item);

           if (trailersKeys != null && trailersKeys.size() > 0) {
               String url = "https://www.youtube.com/watch?v=" + trailersKeys.get(0);

               shareIntent = new Intent(android.content.Intent.ACTION_SEND);
               shareIntent.setType("text/plain");
               shareIntent.putExtra(Intent.EXTRA_STREAM, url);
               shareIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET);
           }
           if (mShareActionProvider != null) {
               mShareActionProvider.setShareIntent(shareIntent);
           }
       }

        return super.onOptionsItemSelected(item);
    }


}

File: app/src/main/java/com/udacity/google/popularmovies/fragments/DetailFragment.kt
package com.udacity.google.popularmovies.fragments

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.view.MenuItemCompat
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.ListView
import android.support.v7.widget.ShareActionProvider
import android.widget.TextView
import android.widget.Toast
import android.widget.ToggleButton

import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.squareup.picasso.Picasso
import com.udacity.google.popularmovies.util.FavouriteDataSource
import com.udacity.google.popularmovies.R

import org.json.JSONArray
import org.json.JSONException

import java.util.ArrayList

class DetailFragment : Fragment() {

    private var clickedMovieId: Int = 0
    private lateinit var reviewsLV: ListView
    private lateinit var trailersLv: ListView
    private lateinit var queue: RequestQueue
    private lateinit var trailersNames: ArrayList<String>
    private var trailersKeys: ArrayList<String>? = null
    private var mShareActionProvider: ShareActionProvider? = null
    internal lateinit var shareIntent: Intent

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val rootView = inflater.inflate(R.layout.detail_fragment, null, false)
        val favButton = rootView.findViewById<View>(R.id.FavtoggleButton) as ToggleButton
        clickedMovieId = MainFragment.clickedMovie.id
        val movieTitle = rootView.findViewById<View>(R.id.movie_title) as TextView
        movieTitle.text = MainFragment.clickedMovie.title
        val moviePoster = rootView.findViewById<View>(R.id.movie_poster) as ImageView
        Picasso.with(activity).load(MainFragment.clickedMovie.poster_path).resize(300, 300).centerInside().into(moviePoster)
        Log.d("DetailDebug", MainFragment.clickedMovie.poster_path)
        val movieDate = rootView.findViewById<View>(R.id.movie_date) as TextView
        movieDate.text = MainFragment.clickedMovie.release_date

        val movieVoteAverage = rootView.findViewById<View>(R.id.movie_vote_average) as TextView
        movieVoteAverage.text = MainFragment.clickedMovie.vote_average

        val movieOverView = rootView.findViewById<View>(R.id.movie_overview) as TextView
        movieOverView.text = MainFragment.clickedMovie.overview

        favButton.setOnCheckedChangeListener { compoundButton, b ->
            val favouriteDataSource = FavouriteDataSource(activity!!)
            favouriteDataSource.open()

            if (favButton.text == "MARK AS A FAVOURITE") {
                //                    save clicked movie id and poster link to db
                favouriteDataSource.create(MainFragment.clickedMovie)
                Toast.makeText(activity, "Marked as FAVOURITE", Toast.LENGTH_LONG).show()
            } else {
                favouriteDataSource.delete(MainFragment.clickedMovie.id)
                Toast.makeText(activity, "Marked as not favourite", Toast.LENGTH_LONG).show()
            }

            favouriteDataSource.close()
        }

        trailersLv = rootView.findViewById<View>(R.id.trailers_list_view) as ListView

        trailersLv.setOnTouchListener { v, event ->
            // Setting on Touch Listener for handling the touch inside ScrollView
            // Disallow the touch request for parent scroll on touch of child view
            v.parent.requestDisallowInterceptTouchEvent(true)
            false
        }


        queue = Volley.newRequestQueue(activity!!)


        trailersNames = ArrayList()
        trailersKeys = ArrayList()

        var builder = Uri.Builder()

        builder.scheme("http")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath("" + clickedMovieId)
                .appendPath("videos")
                .appendQueryParameter("api_key", activity!!.resources.getString(R.string.api_key))

        //http://api.themoviedb.org/3/movie/49026/videos?api_key=4f92be250f018aff8f3a2b5c3864aecd
        var myUrl = builder.build().toString()

        Log.d("VolleyDebug", myUrl)
        val trailersRequest = JsonObjectRequest(Request.Method.GET, myUrl, null, Response.Listener { response ->
            var resultJsonArray: JSONArray? = null
            try {

                resultJsonArray = response.getJSONArray("results")


                for (i in 0 until resultJsonArray!!.length()) {

                    val jsonObject = resultJsonArray.getJSONObject(i)

                    val name = jsonObject.getString("name")
                    val key = jsonObject.getString("key")

                    trailersNames.add(name)
                    trailersKeys!!.add(key)
                    val adapter = CustomListAdapter(activity!!, trailersNames)

                    trailersLv.adapter = adapter

                }

            } catch (e: JSONException) {
                e.printStackTrace()
            }
        },
                Response.ErrorListener { Toast.makeText(activity, "Sorry some thing happened while retrieving trailers, please try again", Toast.LENGTH_LONG).show() }
        )

        queue.add(trailersRequest)


        trailersLv.onItemClickListener = AdapterView.OnItemClickListener { adapterView, view, i, l ->
            val clickedKey = trailersKeys!![i]

            val url = "https://www.youtube.com/watch?v=$clickedKey"
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = Uri.parse(url)
            startActivity(intent)
        }

        reviewsLV = rootView.findViewById<View>(R.id.reviews_list_view) as ListView

        val reviewsArrayList = ArrayList<String>()


        builder = Uri.Builder()

        builder.scheme("http")
                .authority("api.themoviedb.org")
                .appendPath("3")
                .appendPath("movie")
                .appendPath("" + clickedMovieId)
                .appendPath("reviews")
                .appendQueryParameter("api_key", activity!!.resources.getString(R.string.api_key))

        //http://api.themoviedb.org/3/movie/49026/reviews?api_key=4f92be250f018aff8f3a2b5c3864aecd

        myUrl = builder.build().toString()


        val reviewsRequest = JsonObjectRequest(Request.Method.GET, myUrl, null, Response.Listener { response ->
            try {
                val resultsJsonArray = response.getJSONArray("results")


                for (i in 0 until resultsJsonArray.length()) {

                    val resultJsonObj = resultsJsonArray.getJSONObject(i)

                    val author = resultJsonObj.getString("author")

                    val content = resultJsonObj.getString("content")

                    //TODO: author with content !
                    reviewsArrayList.add(author + ": " + System.getProperty("line.separator") + System.getProperty("line.separator") + content)

                    Log.d("DetailDebug", "$author: /r/n$content")

                    val reviewsAdapter = ArrayAdapter(activity!!, android.R.layout.simple_list_item_1, reviewsArrayList)
                    reviewsLV.adapter = reviewsAdapter

                }

            } catch (e: JSONException) {
                e.printStackTrace()
            }
        },
                Response.ErrorListener { }
        )

        queue.add(reviewsRequest)


        reviewsLV.setOnTouchListener { v, event ->
            // Setting on Touch Listener for handling the touch inside ScrollView
            // Disallow the touch request for parent scroll on touch of child view
            v.parent.requestDisallowInterceptTouchEvent(true)
            false
        }


        val favouriteDataSource = FavouriteDataSource(activity!!)
        favouriteDataSource.open()
        if (favouriteDataSource.isFavouriteMovie(clickedMovieId)) {

            favButton.text = "FAVOURITE"
        }


        return rootView
    }

    inner class CustomListAdapter(//ArrayList<String> images ;
            internal var context: Activity, internal var stringArrayList: ArrayList<String>) : ArrayAdapter<Any>(context, R.layout.grid_view_item) {

        override fun getCount(): Int {
            return stringArrayList.size
        }

        override fun getItem(position: Int): Any? {
            return stringArrayList[position]
        }

        override fun getItemId(position: Int): Long {
            return 0
        }

        override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {

            val inflater = context.layoutInflater

            val rowView = inflater.inflate(R.layout.list_view_item, null, true)

            val tv = rowView.findViewById<View>(R.id.list_item_text_view) as TextView

            tv.text = stringArrayList[position]

            return rowView


        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setHasOptionsMenu(true)

    }


    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {

        inflater!!.inflate(R.menu.menu_detail, menu)


    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {

        if (item!!.itemId == R.id.menu_item_share) {
            mShareActionProvider = MenuItemCompat.getActionProvider(item) as ShareActionProvider

            if (trailersKeys != null && trailersKeys!!.size > 0) {
                val url = "https://www.youtube.com/watch?v=" + trailersKeys!![0]

                shareIntent = Intent(android.content.Intent.ACTION_SEND)
                shareIntent.type = "text/plain"
                shareIntent.putExtra(Intent.EXTRA_STREAM, url)
                shareIntent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_WHEN_TASK_RESET)
            }
            if (mShareActionProvider != null) {
                mShareActionProvider!!.setShareIntent(shareIntent)
            }
        }

        return super.onOptionsItemSelected(item)
    }


}
